packages/truffle-compile/index.js

```diff
@@ -280,72 +280,39 @@ function replaceLinkReferences(bytecode, linkReferences, libraryName) {
   return bytecode;
 }
 
-function orderABI(contract) {
-  var contract_definition;
-  var ordered_function_names = [];
-
-  for (var i = 0; i < contract.legacyAST.children.length; i++) {
-    var definition = contract.legacyAST.children[i];
-
-    // AST can have multiple contract definitions, make sure we have the
-    // one that matches our contract
-    if (
-      definition.name !== "ContractDefinition" ||
-      definition.attributes.name !== contract.contract_name
-    ) {
-      continue;
-    }
-
-    contract_definition = definition;
-    break;
+function orderABI({ abi, contract_name: contractName, ast }) {
+  // AST can have multiple contract definitions, make sure we have the
+  // one that matches our contract
+  const contractDefinition = ast.nodes.filter(
+    ({ nodeType, name }) =>
+      nodeType === "ContractDefinition" && name === contractName
+  )[0];
+
+  if (!contractDefinition || !contractDefinition.nodes) {
+    return abi;
   }
 
-  if (!contract_definition) return contract.abi;
-  if (!contract_definition.children) return contract.abi;
-
-  contract_definition.children.forEach(function(child) {
-    if (child.name === "FunctionDefinition") {
-      ordered_function_names.push(child.attributes.name);
-    }
-  });
+  // Find all function definitions
+  const orderedFunctionNames = contractDefinition.nodes
+    .filter(({ nodeType }) => nodeType === "FunctionDefinition")
+    .map(({ name: functionName }) => functionName);
 
   // Put function names in a hash with their order, lowest first, for speed.
-  var functions_to_remove = ordered_function_names.reduce(function(
-    obj,
-    value,
-    index
-  ) {
-    obj[value] = index;
-    return obj;
-  },
-  {});
-
-  // Filter out functions from the abi
-  var function_definitions = contract.abi.filter(function(item) {
-    return functions_to_remove[item.name] !== undefined;
-  });
-
-  // Sort removed function defintions
-  function_definitions = function_definitions.sort(function(item_a, item_b) {
-    var a = functions_to_remove[item_a.name];
-    var b = functions_to_remove[item_b.name];
-
-    if (a > b) return 1;
-    if (a < b) return -1;
-    return 0;
-  });
-
-  // Create a new ABI, placing ordered functions at the end.
-  var newABI = [];
-  contract.abi.forEach(function(item) {
-    if (functions_to_remove[item.name] !== undefined) return;
-    newABI.push(item);
-  });
-
-  // Now pop the ordered functions definitions on to the end of the abi..
-  Array.prototype.push.apply(newABI, function_definitions);
-
-  return newABI;
+  const functionIndexes = orderedFunctionNames
+    .map((functionName, index) => ({ [functionName]: index }))
+    .reduce((a, b) => Object.assign({}, a, b), {});
+
+  // Construct new ABI with functions at the end in source order
+  return [
+    ...abi.filter(({ name }) => functionIndexes[name] === undefined),
```

Comment:
I gather it's intentional that this sorts events to the top, in arbitrary order.  I'm guessing it's also intentional that it sorts inherited-and-not-overridden functions to the top in arbitrary order as well?  (Given as that's what the old code also did.)

Reply:
This is just how it's been implemented. I'm not sure if that's necessary; I'll find out.

Reply:
Doesn't look like this matters: at least, in the Solidity testing stuff, it just filters out non-functions: https://github.com/trufflesuite/truffle/blob/develop/packages/truffle-core/lib/testing/soliditytest.js#L77

Reply:
(And same as above: I move that we don't bother with this kind of semantic changes until the aforementioned `wire-decoder` stuff is available)